### ISSUE TITLE
twisted's SIGCHLD handler breaks popen.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10460]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#733|
|Type|defect|
|Created|2004-09-30 06:57:33Z|
|Branch|https://github.com/twisted/twisted/tree/popen-compatibility-733-6|

Highest priority is for regressions and reviews.

<details><summary>Searchable metadata</summary>

```
trac-id__733 733
type__defect defect
reporter__jknight jknight
priority__high high
milestone__ 
branch__branches_popen_compatibility_733_6 branches/popen-compatibility-733-6
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1096527453000000 1096527453000000
changetime__1266899432000000 1266899432000000
version__ 
owner__ 
cc__glyph cc__exarkun cc__spiv cc__itamarst cc__jknight cc__titty cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This is a simple way to demonstrate:
>>> from twisted.internet import reactor
>>> reactor._handleSignals()
>>> f=os.popen("sleep 5; echo 'Foo'"); f.read()
Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
IOError: [Errno 4] Interrupted system call

The solution is easy: our SIGCHLD handler needs to be given a SA_RESTART flag. This will cause it to 
not be handled until after the read call returns, as python does not allow the python sighandler code to 
be run in the interrupt context. But that's okay.

There's only one minor problem: the python signal() wrapper doesn't provide the flags parameter that's 
necessary in order to specify SA_RESTART. So, to fix this, someone should write a C extension module 
that does something like this:

void add_signalflags(int sig, int moreflags)
{
  int res;
  struct sigaction act;
  res = sigaction(sig, NULL, &amp;act);
  if (res &lt; 0)
    RAISE ERROR
  
  act.sa_flags |= moreflags;
  res = sigaction(sig, &amp;act, NULL);
  if (res &lt; 0)
    RAISE ERROR
}
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
OK, no useful feedback, but AWESOME for figuring this out, james.  Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
OK wait no, I changed my mind.  Will passing this flag still (properly)
interrupt select()?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
According to my documentation it will.
     If a signal is caught during the system calls listed below, the call may
     be forced to terminate with the error EINTR, the call may return with a
     data transfer shorter than requested, or the call may be restarted.
     Restart of pending calls is requested by setting the SA_RESTART bit in
     sa_flags.  The affected system calls include open(2), read(2), write(2),
     sendto(2), recvfrom(2), sendmsg(2) and recvmsg(2) on a communications
     channel or a slow device (such as a terminal, but not a regular file) and
     during a wait(2) or ioctl(2).  However, calls that have already committed
     are not restarted, but instead return a partial success (for example, a
     short read count).

May want to make *sure* by testing, though.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
On Redhat 7.3 I had a hard time reproducing this, but:

>>> signal.alarm(1); f=os.popen("sleep 5; echo 'Foo'"); f.read()

reproduced it some of the time.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="ColinAlston's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ColinAlston commented|
|-|-|

Reproduced thanks to Gentoo Emerge and its crappy signaling

 avail = os.popen('emerge -p  --nocolor --nospinner %s | grep %s' % (pack, pack))
 k = avail.read()

Something like that, put it anywhere you please, and you will intermitantly generate this bug on Twisted 2.0.1

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

moshez made an interesting suggestion on IRC -- I wonder if we can implement the missing add_signalflags function with ctypes?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

See also [#10513](https://github.com/twisted/twisted/issues/10513) and [#1997](https://github.com/twisted/twisted/issues/1997)

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|titty commented|
|-|-|

Feel free to use: http://systemexit.de/repo/sigalter
It's a c module, which provides a set_SA_RESTART(SIGNUM) method.

Here's the code (in case the above server is unreachable):


```
#include <Python.h>
#include <signal.h>

static PyObject *
set_SA_RESTART (void *self, PyObject *args)
{
        struct sigaction sa;

        int signum;
        if (!PyArg_ParseTuple(args, "i:set_SA_RESTART", &signum)) {
                return NULL;
        }

        if (sigaction(signum, NULL, &sa)==-1) {
                goto error;
        }

        if (!(sa.sa_flags & SA_RESTART)) {
                sa.sa_flags |= SA_RESTART;
                if (sigaction(signum, &sa, NULL)==-1) {
                        goto error;
                }
        }

        return Py_BuildValue("");

error:
        PyErr_SetFromErrno(PyExc_OSError);
        return NULL;
}

static PyMethodDef module_function[] = {
        {"set_SA_RESTART", (PyCFunction)set_SA_RESTART, METH_VARARGS, 
         "set_SA_RESTART(SIGNUM):\n   sets SA_RESTART flag on the signalhandler for signal SIGNUM"},
        { NULL,    NULL }
};


static char module_doc[] = "provide funtions to alter signal handling";

DL_EXPORT(void) initsigalter(void)
{
        Py_InitModule3("sigalter", module_function, module_doc);
}
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|titty commented|
|-|-|

the ctypes version (for linux):
```
import ctypes
import signal
libc = ctypes.CDLL("libc.so.6")
libc.siginterrupt(signal.SIGCHLD, 0)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

[#10513](https://github.com/twisted/twisted/issues/10513), [#1997](https://github.com/twisted/twisted/issues/1997), and [[[#2535](https://github.com/twisted/twisted/issues/2535)](https://github.com/twisted/twisted/issues/2535)](https://github.com/twisted/twisted/issues/2535) have significant overlap with this.  Fixing any one of these (except for #2535, perhaps) almost certainly means fixing all of them.

The correct fix for most of them is:

  * Write a SIGCHLD handler in C.  All it needs to do is write to an fd.
  * Install the SIGCHLD handler with the SA_RESTART flag.  This makes it run when a child exits, but doesn't cause EINTR.
  * Include the fd in the reactor so that whenever it is written to, the reactor will wake up and reap dead children.

With this solution:

  * No more EINTR
  * No more race condition inside ceval.c

However, a custom SIGCHLD handler is still required, so #2535 isn't resolved.

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|titty commented|
|-|-|

http://docs.python.org/dev/library/signal.html:

python 2.6 will have signal.siginterrupt and signal.set_wakeup_fd.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra commented|
|-|-|

Replying to [titty](#note_14):
> http://docs.python.org/dev/library/signal.html:
>
> python 2.6 will have signal.siginterrupt and signal.set_wakeup_fd.

Now a Python 2.6 buildslave available http://buildbot.twistedmatrix.com/builders/ubuntu64-python2.6-select/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [25649]) Branching to 'popen-compatibility-733'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Should I kill off the branches for [#10513](https://github.com/twisted/twisted/issues/10513)?  It looks like the code useful for addressing this problem has all been inherited by this branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Also if you're working on a ticket, please accept it (or at least reassign it to yourself).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

Deleting branches is so boring.  I think I may have gotten all the good bits of the [#10513](https://github.com/twisted/twisted/issues/10513) branch at this point, so feel free to delete it if you like.  Someday I'll probably write a program that deletes branches which are associated with tickets which have been closed for a while.  Until then I don't plan to think about it much.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [26139]) Branching to 'popen-compatibility-733-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

This is weird.

Under 2.5, using many different variations of James's example, I can no longer reproduce this problem.  It still raises an exception on 2.5 though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

Ugh, hit "submit" too early.

Under _python_ 2.5 _and 2.6_, and Twisted trunk, using many different variations of James's example, I can no longer reproduce this problem. It still raises an exception on _python 2.4_ though.

I can't see what's changed.  `PyOS_setsig` seems to be the same, at least in 2.5.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_24"></a>|@jyknight commented|
|-|-|

So what's left to do on that branch?

From a cusory review, I'd say: 
 * it needs a test.
 * It shouldn't use the custom _sigchld.c module under python 2.6.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra commented|
|-|-|

Not able to reproduce this either on osx 10.5.8:

```python
Python 2.5.1 (r251:54863, Feb  6 2009, 19:02:12) 
[GCC 4.0.1 (Apple Inc. build 5465)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> from twisted.internet import reactor
>>> f = os.popen("sleep 5; echo 'Foo'"); f.read()
'Foo\n'
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

It's a real issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [27731]) Branching to 'popen-compatibility-733-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [28250]) Branching to 'popen-compatibility-733-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun removed owner|
|-|-|

Okay!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_30"></a>|@dreid set owner to @exarkun|
|-|-|

 * Lots of undefined references (#NNNN)
 * twisted/internet/test/test_baseprocess.py has it's copyright line updated but no changes
 * There isn't a test showing that twisted.internet.posixbase._UnixWaker.wakeUp raises when the write fails with something other than EAGAIN (It's in the diff but only because the classes position shifted in the file)
 * Possible typo in PosixReactorBase ```mopde = self._checkMode('IReactorUNIXDatagram.connectUNIXDatagram', mode)```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

(In [28302]) Branching to 'popen-compatibility-733-5'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun removed owner|
|-|-|

Thanks.  Fixed those points.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_34"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

So, uh... oops?  I wanted to see a all-supported-builders build report for this branch, so I did this:

```
glyph@miranda:~/Projects/Twisted/sandbox/exarkun$ python force-builds.py 
Traceback (most recent call last):
  File "force-builds.py", line 10, in <module>
    from twisted.internet import reactor, defer
  File "/Domicile/glyph/Projects/Twisted/trunk/twisted/internet/reactor.py", line 37, in <module>
    from twisted.internet import selectreactor
  File "/Domicile/glyph/Projects/Twisted/trunk/twisted/internet/selectreactor.py", line 21, in <module>
    from twisted.internet import posixbase
  File "/Domicile/glyph/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 181, in <module>
    _Waker = _PipeWaker
NameError: name '_PipeWaker' is not defined
[Error: 1]
```
but I think that's the review by itself.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [28362]) Update to use the new name

refs [#10460](https://github.com/twisted/twisted/issues/10460)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun removed owner|
|-|-|

Sigh, forgot about that.  Fixed now.

**Comment by twisted-trac** [comment]
|<img alt="terrycojones's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_37"></a>|terrycojones commented|
|-|-|

The example shown at the top of the ticket works for me. trial twisted reports no errors.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

(In [28435]) Branching to 'popen-compatibility-733-6'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

(In [28445]) Quiet the error about the result of `write` being ignored

refs [#10460](https://github.com/twisted/twisted/issues/10460)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_41"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Yay for C code!

It appears that the feedback from other reviewers has been adequately dealt with.  I can't find anything else to complain about.

Merge, and rejoice.  (In that order.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun set status to `closed`|
|-|-|

(In [28447]) Merge popen-compatibility-733-6

Author: glyph, itamar, exarkun, (perhaps others sorry if I forgot you)
Reviewer: glyph, dreid, jknight
Fixes: [#10460](https://github.com/twisted/twisted/issues/10460)

Change the reactors to, on POSIX, use `signal.set_wakeup_fd`
and `signal.siginterrupt` to support child processes without
creating the possibility for causing syscalls to fail with
EINTR.  On Python 2.5 and older, where these APIs are not
available, a new extension module is added to implement the
functionality.  If neither of these is available, revert to
the old behavior.  Also Glib2/Gtk2 reactor retain the old
behavior due to implementation difficulties.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
